build-api/templates/dockerfile.tmplt

```diff
@@ -6,7 +6,7 @@ RUN eval `ssh-agent -s` && \
     git clone -b {{ .BRANCH }} git@github.com:{{ .REPO_OWNER }}/{{ .REPO_NAME }}.git
 
 RUN cd {{ .REPO_NAME }} && git pull origin {{ .BRANCH }}
-RUN cd {{ .REPO_NAME }} && git checkout {{ .HASH }}
+RUN cd {{ .REPO_NAME }} && {{if .TAG}} git checkout -b {{ .TAG }} {{else}} git checkout {{ .HASH }} {{end}}
```

Comment:
I think we could also do most of this on line 6. `git clone -b tag/branch` (cloning treats branches and tags the same way. `{{if not .TAG}} git checkout hash`
Doing that, I think, carries the added benefit of downloading a smaller payload from github. 
http://stackoverflow.com/a/27421557

Also, apparently git NOW supports cloning a specific commit hash. That would be HUGE performance wise. 
http://stackoverflow.com/questions/14872486/retrieve-specific-commit-from-a-remote-git-repository/30701724#30701724

Perhaps just open a separate PR for this optimization.